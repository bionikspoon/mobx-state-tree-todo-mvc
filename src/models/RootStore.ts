import { types, Instance, cast } from "mobx-state-tree"
import { v4 as uuid } from "uuid"

const Todo = types
  .model("Todo", {
    id: types.optional(types.identifier, () => uuid()),
    label: types.string,
    completed: false,
  })
  .actions(self => ({
    toggle(completed = !self.completed) {
      self.completed = completed
    },
    setLabel(label: string) {
      self.label = label
    },
  }))

export const RootStore = types
  .model("Root", {
    todos: types.optional(types.array(Todo), []),
    editing: "",
  })
  .views(self => ({
    get activeTodos() {
      return self.todos.filter(todo => !todo.completed)
    },
    get completedTodos() {
      return self.todos.filter(todo => todo.completed)
    },
    get allCompleted() {
      return self.todos.every(todo => todo.completed)
    },
  }))
  .views(self => ({
    get activeTodosCount() {
      return self.activeTodos.length
    },
  }))
  .actions(self => ({
    addTodo(name: string, completed = false) {
      self.todos.push(Todo.create({ label: name, completed: completed }))
    },
    editTodo(id: string) {
      self.editing = id
    },
    removeTodo(id: string) {
      self.todos = cast(self.todos.filter(todo => todo.id !== id))
    },
    clearCompletedTodos() {
      self.todos = cast(self.todos.filter(todo => !todo.completed))
    },
    toggleAll() {
      const completed = !self.allCompleted
      self.todos.forEach(todo => todo.toggle(completed))
    },
  }))

export type RootInstance = Instance<typeof RootStore>
