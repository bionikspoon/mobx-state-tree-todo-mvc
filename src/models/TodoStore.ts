import { types, Instance, destroy, getParentOfType } from "mobx-state-tree"
import { v4 as uuid } from "uuid"

const Todo = types
  .model("Todo", {
    id: types.optional(types.identifier, () => uuid()),
    label: types.string,
    completed: false,
  })
  .actions(self => ({
    toggle(completed = !self.completed) {
      self.completed = completed
    },
    setLabel(label: string) {
      self.label = label
    },
    remove() {
      getParentOfType(self, TodoStore).removeTodo(self as TodoInstance)
    },
  }))
export type TodoInstance = Instance<typeof Todo>

export const TodoStore = types
  .model("Root", {
    todos: types.optional(types.array(Todo), []),
    editing: "",
  })
  .views(self => ({
    get activeTodos() {
      return self.todos.filter(todo => !todo.completed)
    },
    get completedTodos() {
      return self.todos.filter(todo => todo.completed)
    },
    get allCompleted() {
      return self.todos.every(todo => todo.completed)
    },
  }))
  .views(self => ({
    get activeTodosCount() {
      return self.activeTodos.length
    },
  }))
  .actions(self => ({
    addTodo(name: string, completed = false) {
      self.todos.push(Todo.create({ label: name, completed: completed }))
    },
    editTodo(id: string) {
      self.editing = id
    },
    removeTodo(todo: TodoInstance) {
      destroy(todo)
    },
    clearCompletedTodos() {
      self.todos.replace(self.todos.filter(todo => !todo.completed))
    },
    toggleAll() {
      const completed = !self.allCompleted
      self.todos.forEach(todo => todo.toggle(completed))
    },
  }))

export type TodoStoreInstance = Instance<typeof TodoStore>
