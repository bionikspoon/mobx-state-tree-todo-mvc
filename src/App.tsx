import React, { ReactElement } from "react"

import { RootStore } from "./models/RootStore"
import TodoApp from "./TodoApp"

const store = RootStore.create()

store.addTodo("Taste JavaScript", true)
store.addTodo("Buy a unicorn")

declare global {
  interface Window {
    store: typeof store
  }
}
window.store = store

interface Props {}

export default function App(_props: Props): ReactElement {
  return <TodoApp store={store} />
}
