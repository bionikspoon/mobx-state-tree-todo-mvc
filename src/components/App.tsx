import React from "react"
import { BrowserRouter as Router, Route } from "react-router-dom"

import { TodoStoreInstance } from "../models/TodoStore"
import TodoApp from "./TodoApp"
import Footer from "./Footer"

interface AppProps {
  store: TodoStoreInstance
}

const App: React.FC<AppProps> = props => (
  <>
    <Router>
      <Route path="/:view?">
        <TodoApp store={props.store} />
      </Route>
    </Router>
    <Footer />
  </>
)

export default App
