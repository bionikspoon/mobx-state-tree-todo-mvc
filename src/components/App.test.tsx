import React from "react"
import { render } from "@testing-library/react"
import App from "./App"
import { TodoStore } from "../models/TodoStore"

describe("<App />", () => {
  const state = {
    todos: [
      {
        label: "Taste JavaScript",
        completed: true,
      },
      {
        label: "Buy a unicorn",
        completed: false,
      },
    ],
  }
  test("render", () => {
    const { asFragment } = render(<App store={TodoStore.create(state)} />)

    expect(asFragment()).toMatchSnapshot()
  })

  describe("when there are no todos", () => {
    const store = { todos: [] }

    test.each`
      text
      ${"All"}
      ${"Active"}
      ${"Completed"}
    `("should hide the #footer $text link", ({ text }) => {
      const { queryByText } = render(<App store={TodoStore.create(store)} />)

      expect(queryByText(text, { selector: "a" })).not.toBeInTheDocument()
    })

    it("should hide #main", () => {
      const { queryByLabelText } = render(
        <App store={TodoStore.create(store)} />
      )

      expect(queryByLabelText("Mark all as complete")).not.toBeInTheDocument()
    })
  })
})
