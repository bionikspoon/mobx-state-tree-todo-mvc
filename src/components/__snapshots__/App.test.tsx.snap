// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> render 1`] = `
<DocumentFragment>
  <section
    class="todoapp"
  >
    <header
      class="header"
    >
      <h1>
        todos
      </h1>
      <input
        class="new-todo"
        placeholder="What needs to be done?"
        type="text"
        value=""
      />
    </header>
    <section
      class="main"
    >
      <input
        class="toggle-all"
        id="toggle-all"
        type="checkbox"
      />
      <label
        for="toggle-all"
      >
        Mark all as complete
      </label>
      <ul
        class="todo-list"
      >
        <li
          class="completed"
        >
          <div
            class="view"
          >
            <input
              checked=""
              class="toggle"
              type="checkbox"
            />
            <label>
              Taste JavaScript
            </label>
            <button
              class="destroy"
            />
          </div>
        </li>
        <li
          class=""
        >
          <div
            class="view"
          >
            <input
              class="toggle"
              type="checkbox"
            />
            <label>
              Buy a unicorn
            </label>
            <button
              class="destroy"
            />
          </div>
        </li>
      </ul>
    </section>
    <footer
      class="footer"
    >
      <span
        class="todo-count"
      >
        <strong>
          1
        </strong>
         item left
      </span>
      <ul
        class="filters"
      >
        <li>
          <a
            aria-current="page"
            class="selected"
            href="/"
          >
            All
          </a>
        </li>
        <li>
          <a
            href="/active"
          >
            Active
          </a>
        </li>
        <li>
          <a
            href="/completed"
          >
            Completed
          </a>
        </li>
      </ul>
      <button
        class="clear-completed"
      >
        Clear completed
      </button>
    </footer>
  </section>
  <footer
    class="info"
  >
    <p>
      Double-click to edit a todo
    </p>
    <p>
      Created by 
      <a
        href="https://manuphatak.com"
      >
        Manu Phatak
      </a>
    </p>
    <p>
      Part of 
      <a
        href="http://todomvc.com"
      >
        TodoMVC
      </a>
    </p>
  </footer>
</DocumentFragment>
`;
